### ISSUE TITLE
Why is a zero array concatenated to create shadow labels, features and coordinates?

### ISSUE THREAD

**Issue by sandeepnmenon** [MCG-NJU/CGA-Net#1]
As shown in lines
https://github.com/MCG-NJU/CGA-Net/blob/84aef058dd6c56919e37375e1b8880605094e6a4/CGA/cga.py#L84
https://github.com/MCG-NJU/CGA-Net/blob/84aef058dd6c56919e37375e1b8880605094e6a4/CGA/cga.py#L89
https://github.com/MCG-NJU/CGA-Net/blob/84aef058dd6c56919e37375e1b8880605094e6a4/CGA/cga.py#L166

To take an example for shadow_features creation, if
```
features = tensor([[ 1.7550,  0.4128, -0.1563],
        [-0.7864, -0.5018, -1.1231],
        [ 0.6667,  2.6513, -0.0978],
        [-0.8356, -0.7712, -0.7236],
        [-0.4006, -0.2409,  0.3513]])
shadow_features = torch.cat([features, torch.zeros_like(features[:1, :])], dim=0)
```
results in
shadow_features
```
tensor([[ 1.7550,  0.4128, -0.1563],
        [-0.7864, -0.5018, -1.1231],
        [ 0.6667,  2.6513, -0.0978],
        [-0.8356, -0.7712, -0.7236],
        [-0.4006, -0.2409,  0.3513],
        [ 0.0000,  0.0000,  0.0000]])
```

Why is this operation required before gathering the respective values for the neighbor indices?

Thank you

**Comment by inspirelt** [comment]
Hi, @sandeepnmenon. These shadow operations are paddings for the variant-length point clouds, which avoid accessing an illegal address. I've tried to remove them while got a random error in CloserLook3D. You can refer to this https://github.com/HuguesTHOMAS/KPConv/issues/55#issuecomment-571656633. Thank you!

**Comment by sandeepnmenon** [comment]
Thank you @inspirelt, In that case, (if a zero feature is added), then shouldn' the dimenions be one instead of zero?
In that case the shadow features will be like this right? With the zero column.
```
tensor([[ 1.7550,  0.4128, -0.1563],
        [-0.7864, -0.5018, -1.1231, 0,0000],
        [ 0.6667,  2.6513, -0.0978, 0,0000],
        [-0.8356, -0.7712, -0.7236, 0,0000],
        [-0.4006, -0.2409,  0.3513, 0,0000]
])
```
Or is it the one mentioned in the body of the comment?

**Comment by inspirelt** [comment]
The expected shadow features are 
```
tensor([[ 1.7550,  0.4128, -0.1563], # f_0
        [-0.7864, -0.5018, -1.1231], # f_1
        [ 0.6667,  2.6513, -0.0978], # f_2
......
        [-0.8356, -0.7712, -0.7236], # f_{n-2}
        [-0.4006, -0.2409,  0.3513], # f_{n-1}
        [0.0000, 0.0000, 0.0000] # f_n, this is the padding
])
```

So the paddings dimension is zero.

**Comment by sandeepnmenon** [comment]
Thank you @inspirelt for clarifying.

**Comment by whuhxb** [comment]
Hi, @sandeepnmenon 
Have you run the CGA-Net with RandLA-Net on S3DIS?

**Comment by sandeepnmenon** [comment]
@whuhxb No
I ran CGGA-NET with SPVCNN on a privately collected dataset.

**Comment by whuhxb** [comment]
@sandeepnmenon 
OK. How is your performance improvement?

**Comment by whuhxb** [comment]
@sandeepnmenon 
OK. I have questions to ask you. I will write you an e-mail. Thanks.
